### PR TITLE
Added initial examples

### PR DESCRIPTION
**Pull request by turneram**

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+# AMDMIGraphX Usage Examples
```
Comment (scxiao) [inline examples/README.md:1]: Can we can add some introduction words here, then gave a link to the readme in each of the subfolder?

Comment (scxiao) [reply examples/README.md:1]: Maybe we can also add an example using our C/C++ API interface. I attached an example using the C/C++ here. 
[mgx_c_api.tar.gz](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/files/5446293/mgx_c_api.tar.gz)

File: examples/migraphx_driver/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# MIGraphX Driver
+There is a new driver as /opt/rocm/bin/migraphx-driver.  The --help command shows options available
```
Comment (kahmed10) [inline examples/migraphx_driver/README.md:2]:
```suggestion
The MIGraphX driver is located under /opt/rocm/bin/migraphx-driver.  The --help command shows options available
```

File: examples/migraphx_driver/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    perf
+    params
+    read
+    run
+    verify
+    compile
+```
+For example, the read command will read a file and print the internal graph from MIGraphx
+```
+/opt/rocm/bin/migraphx-driver read --onnx /home/mev/source/migraphx_onnx/torchvision/resnet50i64.onnx
```
Comment (kahmed10) [inline examples/migraphx_driver/README.md:19]: can you add a small snippet (maybe up to 5 lines max) of the output?

File: examples/migraphx_driver/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    verify
+    compile
+```
+For example, the read command will read a file and print the internal graph from MIGraphx
+```
+/opt/rocm/bin/migraphx-driver read --onnx /home/mev/source/migraphx_onnx/torchvision/resnet50i64.onnx 
+```
+Another example, the following command measures performance running an ONNX file using the driver
+```
+/opt/rocm/bin/migraphx-driver perf --onnx /home/mev/source/migraphx_onnx/torchvision/resnet50i64.onnx
```
Comment (kahmed10) [inline examples/migraphx_driver/README.md:23]: can you include the performance summary part of the output?

File: examples/migraphx_driver/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+```
+For example, the read command will read a file and print the internal graph from MIGraphx
+```
+/opt/rocm/bin/migraphx-driver read --onnx /home/mev/source/migraphx_onnx/torchvision/resnet50i64.onnx 
+```
+Another example, the following command measures performance running an ONNX file using the driver
+```
+/opt/rocm/bin/migraphx-driver perf --onnx /home/mev/source/migraphx_onnx/torchvision/resnet50i64.onnx 
+```
+The verify command checks internal consistency once read into MIGraphX
```
Comment (kahmed10) [inline examples/migraphx_driver/README.md:25]:
```suggestion
The verify command is a correctness checker and runs both the reference and GPU implementation to ensure the model outputs are consistent 
```

File: examples/migraphx_driver/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+```
+/opt/rocm/bin/migraphx-driver verify --onnx /home/mev/source/migraphx_onnx/torchvision/resnet50i64.onnx 
+```
```
Comment (kahmed10) [inline examples/migraphx_driver/README.md:28]: can you include a brief summary of the expected output after running (i.e., no errors emitted if verify passes)?

File: examples/python_interface/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        trans = cropped.transpose(2,0,1) # convert HWC to CHW
+
+        # convert to float, normalize and make batch size = 1
+        image = np.ascontiguousarray(
+            np.expand_dims(trans.astype('float32')/256.0,0))
+
+        # display the frame
+        cv2.imshow('frame',cropped)
+
+        migraphx_result = model.run({'0':migraphx.argument(image)})
```
Comment (pfultz2) [inline examples/python_interface/README.md:44]: You should be able to write this as `migraphx_result = model.run({'0':image})`

File: examples/python_interface/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
+## Python Interface
+The Python interface has been simplified so that it is no longer necessary to copy in parameters.  An updated version of the webcam example from <a href="https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/wiki/Getting-started:-using-the-new-features-of-MIGraphX-0.2">MIGraphX examples v0.2</a> is shown below.
+
+A careful comparison with previous example shows it is no longer necessary to allocate the parameters on the GPU or to explicitly copy parameters to the GPU or results from the GPU.  The Python interface handles this by default.
```
Comment (pfultz2) [inline examples/python_interface/README.md:4]: The Python interface doesn't handles this, rather the Python API sets `offload_copy=True` by default when compiling the model(see [`migraphx.compile`](https://rocmsoftwareplatform.github.io/AMDMIGraphX/doc/html/reference/py.html#migraphx.compile)) which will insert copies into the migraphx::program.

File: examples/migraphx_docker/README.md [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# Building the MIGraphX Dockerfile
+
+You can build the provided dockerfile with the following command:
+
+    docker build -f Dockerfile.hip-clang -t migraphx .
+
+To run the docker image:
+
+    docker run --device='/dev/kfd' --device='/dev/dri' -v=`pwd`:/data -w /data --group-add video -it migraphx
```
Comment (scxiao) [inline examples/migraphx_docker/README.md:9]: Are we going to download data for the data folder? If not, I think we should remove the -v='pwd':/data and -w /data from this command.

**Review by scxiao — APPROVED** [review]

File: examples/migraphx_docker/Dockerfile.hip-clang [hunk 10]
```diff
@@ -25,4 +26,4 @@ RUN cd /migraphx && git clone --depth=1 --branch develop https://github.com/ROCm
 RUN cd /migraphx && rbuild package --cxx /opt/rocm-3.7.0/llvm/bin/clang++ -d /migraphx/deps -B /migraphx/build -S /migraphx/src/ -DPYTHON_EXECUTABLE=/usr/bin/python3
 RUN dpkg -i /migraphx/build/*.deb
 
-WORKDIR /migraphx
+WORKDIR /migraphx
```
Comment (kahmed10) [inline examples/migraphx_docker/Dockerfile.hip-clang:29]: can you add a newline here?

**Review by kahmed10 — APPROVED** [review]

File: examples/migraphx_docker/README.md [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# Building the MIGraphX Dockerfile
+
+You can build the provided dockerfile with the following command:
+
+    docker build -f Dockerfile.hip-clang -t migraphx .
```
Comment (scxiao) [inline examples/migraphx_docker/README.md:5]: May be we can remove the dockerfile here, you can just refer to the docker files in the top level directory.

File: examples/migraphx_driver/README.md [hunk 12]
```diff
@@ -0,0 +1,604 @@
+# MIGraphX Driver
+
+## Description
+The MIGraphX driver is a tool that allows you to utilize many of the core functions of MIGraphX without having to write your own program. 
+
+## How to Use this Example
+
+The MIGraphX driver is built alongside MIGraphX and can be found in `/opt/rocm/bin/migraphx-driver`
```
Comment (scxiao) [inline examples/migraphx_driver/README.md:8]: Can you add "or in the folder "build/bin" after building the source code?"

File: examples/README.md [hunk 13]
```diff
@@ -1,13 +1,13 @@
...
-    docker build -f Dockerfile.hip-clang -t migraphx .
-
-To run the docker image:
-
-    docker run --device='/dev/kfd' --device='/dev/dri' -v=`pwd`:/data -w /data --group-add video -it migraphx
+## Description
+This directory contains examples of common use cases for MIGraphX.
+
+## Examples:
+- [C++ Parse, Load, and Save Graph Programs](./parse_load_save)
```
[the remark this answers is not in the document]
Comment (turneram) [reply examples/README.md:7]: Fixed

File: examples/cpp_parse_load_save/README.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+```
+migraphx::program p;
+unsigned batch = 1; //Or read in as argument
+migraphx::onnx_options options;
+options.set_default_dim_value(batch);
+p = parse_onnx(input_file, options);
+```
+
+## Loading
```
[the remark this answers is not in the document]
Comment (turneram) [reply examples/cpp_parse_load_save/README.md:17]: Switched

File: examples/cpp_parse_load_save/README.md [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+```
+migraphx::program p = ... <parsed or loaded program>;
+migraphx_file_options options;
+options.format = "msgpack";
+migraphx::save(p, output_file, options);
+```
+
+
+## Running the Example
+The provided example `parse_load_save.cpp` has these features implemented to allow for comparing outputs.
```
[the remark this answers is not in the document]
Comment (turneram) [reply examples/cpp_parse_load_save/README.md:58]: Added

File: examples/cpp_parse_load_save/README.md [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+migraphx_file_options options;
+options.format = "msgpack";
+migraphx::save(p, output_file, options);
+```
+
+
+## Running the Example
+The provided example `parse_load_save.cpp` has these features implemented to allow for comparing outputs. 
+
+To compile and run the example from this directory:
```
[the remark this answers is not in the document]
Comment (turneram) [reply examples/cpp_parse_load_save/README.md:60]: Added

File: examples/cpp_api_inference/README.md [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+Optionally, graph programs may be quantized to fp16 or int8 precision to improve performance and memory usage. 
+
+##### Floating Point 16-bit Precision
+To quantize using fp16, we simply add the following line:
+```
+migraphx::quantize_fp16(prog);
+```
+
+##### Integer 8-bit Precision
+Int8 quantization requires calibration to accurately map ranges of floating point values onto integer values. This calibration can either be performed explicitly before inference or implicitly during inference.
```
[the remark this answers is not in the document]
Comment (turneram) [reply examples/cpp_api_inference/README.md:41]: Fixed

File: examples/cpp_api_inference/README.md [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+for(auto&& name : param_shapes.names())
+{
+    quant_params.add(name, migraphx::argument(param_shapes[name], calib_dig.data()));
+}
+
+quant_opts.add_calibration_data(quant_params);
+migraphx::quantize_int8(prog, targ, quant_opts);
+```
+
+To calibrate during inference, we simply make the following call:
```
[the remark this answers is not in the document]
Comment (turneram) [reply examples/cpp_api_inference/README.md:60]: Fixed

File: examples/migraphx_docker/README.md (hunk 11 above)
Comment (turneram) [reply examples/migraphx_docker/README.md:5]: Switched

File: examples/migraphx_driver/README.md (hunk 12 above)
Comment (turneram) [reply examples/migraphx_driver/README.md:8]: Added

File: examples/cpp_api_inference/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.5)
+project (CAI)
+
+set (MIGRAPHX_FOLDER AMDMIGraphX)
+set (MIGRAPHX_BUILD ${MIGRAPHX_FOLDER}/build)
```
Comment (pfultz2) [inline examples/cpp_api_inference/CMakeLists.txt:5]: This should call `find_package(migraphx)`

File: examples/cpp_api_inference/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+
+set (CMAKE_CXX_STANDARD 14)
+
+include_directories(../../../${MIGRAPHX_FOLDER}/src/api/include)
+link_directories(../../../${MIGRAPHX_BUILD}/lib)
```
Comment (pfultz2) [inline examples/cpp_api_inference/CMakeLists.txt:10]: The include directories are added with `migraphx_c`.

Comment (pfultz2) [reply examples/cpp_api_inference/CMakeLists.txt:10]: Also, the `link_directories` should not be necessary.

File: examples/cpp_api_inference/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+
+set (mnist_inference mnist_inference.cpp)
+
+message("source file: " mnist_inference.cpp " ---> bin: " mnist_inference)
+add_executable(mnist_inference mnist_inference.cpp)
+target_link_libraries(mnist_inference migraphx_c)
```
Comment (pfultz2) [inline examples/cpp_api_inference/CMakeLists.txt:16]: It actually should be `target_link_libraries(mnist_inference migraphx::migraphx_c)` to make sure it doesnt try to do `-lmigraphx_c`.

File: examples/cpp_parse_load_save/README.md [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+migraphx_file_options options;
+options.format = "json";
+p = migraphx::load(input_file, options);
+```
+
+
+## Running the Example
+The provided example [`parse_load_save.cpp`](./parse_load_save.cpp) has these features implemented to allow for comparing outputs. 
+
+Running this example requires that MIGraphX has already been compiled and build on your system. If you have not yet done so, please refer to the instructions found [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/README.md).
```
Comment (pfultz2) [inline examples/cpp_parse_load_save/README.md:60]: We should assume for the examples that migraphx is already installed.

File: examples/cpp_api_inference/CMakeLists.txt (hunk 19 above)
Comment (pfultz2) [reply examples/cpp_api_inference/CMakeLists.txt:5]: [Here](https://rocmdocs.amd.com/en/latest/Installation_Guide/Using-CMake-with-AMD-ROCm.html) is the guide for using cmake with rocm. We probably need to add the line `list (APPEND CMAKE_PREFIX_PATH /opt/rocm/hip /opt/rocm)` so it works with the examples with the installed migraphx packages.

File: examples/cpp_parse_load_save/README.md (hunk 22 above)
Comment (turneram) [reply examples/cpp_parse_load_save/README.md:60]: Removed

File: examples/cpp_api_inference/CMakeLists.txt (hunk 21 above)
Comment (turneram) [reply examples/cpp_api_inference/CMakeLists.txt:16]: Fixed

File: examples/cpp_api_inference/CMakeLists.txt (hunk 20 above)
Comment (turneram) [reply examples/cpp_api_inference/CMakeLists.txt:10]: Removed

File: examples/cpp_api_inference/CMakeLists.txt (hunk 19 above)
Comment (turneram) [reply examples/cpp_api_inference/CMakeLists.txt:5]: Fixed

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**